specification/dartLangSpec.tex

```diff
@@ -8848,53 +8851,84 @@ \subsubsection{Unqualified Invocation}
-and there is no lexically visible declaration named \id{} in scope.
+For the static analysis of $i$,
+a lexical lookup of \id{} is performed
+(\ref{lexicalLookup}):
 
 \LMHash{}%
-If there exists a lexically visible declaration named \id,
-let $D_{id}$ be the innermost such declaration.
-Then:
+\Case{Lexical lookup yields a declaration}
+In this case the lexical lookup of \id{}
+yields a declaration $D$.
 \begin{itemize}
-\item Consider the situation where $D_{id}$ is a type declaration.
-  If $D_{id}$ is a declaration of a class $C$
+\item
+  Consider the case where $D$ is a type declaration.
+  If $D$ is a declaration of a class $C$
   that has a constructor named $C$
   then the meaning of $i$ depends on the context:
   If $i$ occurs in a constant context
   (\ref{constantContexts}),
-  then $i$ is equivalent to \code{\CONST\,\,$i$};
+  then $i$ is treated as
+  (\ref{notation})
+  \code{\CONST\,\,$i$};
   if $i$ does not occur in a constant context
-  then $i$ is equivalent to \code{\NEW\,\,$i$}.
+  then $i$ is treated as \code{\NEW\,\,$i$}.
   Otherwise a compile-time error occurs
-  (\commentary{that is, if $D_{id}$ does not declare a class,
+  (\commentary{that is, if $D$ does not declare a class,
     or it declares a class that has no constructor named $C$}).
-\item Otherwise, if $D_{id}$ is an import directive
-  where \id{} is declared to be a library prefix,
-  a compile-time error occurs.
-\item Otherwise, if $D_{id}$ declares
+\item
+  Otherwise, if $D$ is a declaration of
   a local function,
   a library function, or
   a library or static getter, or a variable,
-  then $i$ is treated as a function expression invocation
+  then $i$ is treated as
+  (\ref{notation})
+  a function expression invocation
   (\ref{functionExpressionInvocation}).
-\item Otherwise, if $D_{id}$ is a static method of the enclosing class $C$,
-  $i$ is equivalent to
-  \code{$C$.\id<$A_1, \ldots,\ A_r$>($a_1, \ldots,\ a_n,\ x_{n+1}$: $a_{n+1}, \ldots,\ x_{n+k}$: $a_{n+k}$)}.
-\item Otherwise, if $i$ occurs in an instance method body,
-  $i$ is equivalent to the ordinary method invocation
-
-  \code{\THIS{}.\id<$A_1, \ldots,\ A_r$>($a_1, \ldots,\ a_n,\ x_{n+1}$: $a_{n+1}, \ldots,\ x_{n+k}$: $a_{n+k}$)}.
+\item
+  Otherwise, if $D$ is a static method, getter, or variable declaration
+  in the enclosing class $C$,
+  $i$ is treated as
+  (\ref{notation})
+  \code{$C$.$i$}
+  (\ref{ordinaryInvocation}).
+\item
+  \commentary{%
+    A lexical lookup will never yield a declaration $D$
+    which is an instance member.%
+  }
 \end{itemize}
+\vspace{-2ex}
+\EndCase
+
+\Case{Lexical lookup yields an import prefix}
+When the lexical lookup of \id{} yields an import prefix,
+a compile-time error occurs.
+\EndCase
+
+\Case{Lexical lookup yields a member signature}
+When the lexical lookup of \id{} yields a member signature,
+$i$ is treated as
+(\ref{notation})
+the ordinary method invocation
+\code{\THIS.$i$}
+(\ref{ordinaryInvocation}).
+
```

Comment:
(Be prepared for extensions: If $D$ is an extension member invocation, then you can't just rewrite it with a prefix, you have to replace `id` by `Ext<typeArgs>(id)`.)

Reply:
Added a TODO marked 'extension methods'.